File: tests/career-path-service.test.ts
import { describe, expect, it } from "vitest";

import type { CareerPathRepository } from "../src/repositories/career-path-repository";
import type { LearningRepository } from "../src/repositories/learning-repository";
import type { RoleRepository } from "../src/repositories/role-repository";
import {
  buildCareerPathResult,
  CareerPathService,
} from "../src/services/career-path-service";
import type {
  CandidateCareerPath,
  SkillLearningOptions,
} from "../src/types/career-path";
import type {
  RoleRequirementRow,
  RoleSummary,
  SkillRequirement,
} from "../src/types/role";

const roles: Record<string, RoleSummary> = {
  "frontend-developer": {
    id: "frontend-developer",
    slug: "frontend-developer",
    name: "Frontend Developer",
    category: "Web Development",
    seniority: "entry",
    summary: "Builds accessible browser interfaces for product users.",
  },
  "full-stack-developer": {
    id: "full-stack-developer",
    slug: "full-stack-developer",
    name: "Full Stack Developer",
    category: "Web Development",
    seniority: "mid",
    summary: "Builds complete product features across browser and server.",
  },
  "ai-engineer": {
    id: "ai-engineer",
    slug: "ai-engineer",
    name: "AI Engineer",
    category: "AI & Data Science",
    seniority: "senior",
    summary: "Builds evaluated AI features and production model services.",
  },
};

function requirement(
  id: string,
  importance: number,
  essential = true,
): SkillRequirement {
  return {
    id,
    slug: id,
    name: id,
    category: "Test",
    description: `Practical knowledge of ${id} for test scenarios.`,
    importance,
    requiredLevel: "intermediate",
    essential,
  };
}

const requirements = new Map<string, SkillRequirement[]>([
  ["frontend-developer", [requirement("javascript", 5)]],
  [
    "full-stack-developer",
    [requirement("javascript", 5), requirement("node-js", 4)],
  ],
  [
    "ai-engineer",
    [requirement("python", 5), requirement("machine-learning", 4)],
  ],
]);

const learningOptions = new Map<string, SkillLearningOptions>([
  [
    "python",
    {
      skillId: "python",
      resources: [
        {
          id: "python-tutorial",
          title: "The Python Tutorial",
          type: "tutorial",
          provider: "Python Software Foundation",
          url: "https://docs.python.org/3/tutorial/",
          description: "Official Python learning material for language fundamentals.",
        },
      ],
      projects: [],
    },
  ],
]);

const longCandidate: CandidateCareerPath = {
  roles: [
    roles["frontend-developer"],
    roles["full-stack-developer"],
    roles["ai-engineer"],
  ],
  transitions: [
    { difficulty: "moderate", reason: "Expand into server-side product delivery." },
    { difficulty: "hard", reason: "Develop machine-learning and Python capabilities." },
  ],
  hops: 2,
};

const shortCandidate: CandidateCareerPath = {
  roles: [roles["frontend-developer"], roles["ai-engineer"]],
  transitions: [
    { difficulty: "hard", reason: "Build the missing AI engineering foundation." },
  ],
  hops: 1,
};

describe("career path normalization", () => {
  it("calculates shared and missing skills with learning options", () => {
    const result = buildCareerPathResult(
      longCandidate,
      [],
      requirements,
      learningOptions,
    );

    expect(result.suitabilityScore).toBe(28);
    expect(result.steps[0].sharedSkills.map((skill) => skill.id)).toEqual([
      "javascript",
    ]);
    expect(
      result.steps[1].missingEssentialSkills.find(
        (skill) => skill.id === "python",
      )?.resources[0].id,
    ).toBe("python-tutorial");
  });

  it("orders shorter paths before suitability and limits the response", async () => {
    const rows: RoleRequirementRow[] = [...requirements].flatMap(
      ([roleId, roleRequirements]) =>
        roleRequirements.map((roleRequirement) => ({
          roleId,
          requirement: roleRequirement,
        })),
    );
    const roleRepository: RoleRepository = {
      search: async () => [],
      findById: async (id) => roles[id] ?? null,
      findRequirements: async () => rows,
    };
    const careerPathRepository: CareerPathRepository = {
      findCandidates: async () => [longCandidate, shortCandidate],
    };
    const learningRepository: LearningRepository = {
      findOptionsForSkills: async () => [...learningOptions.values()],
    };
    const service = new CareerPathService(
      roleRepository,
      careerPathRepository,
      learningRepository,
    );

    const results = await service.findPaths({
      currentRoleId: "frontend-developer",
      targetRoleId: "ai-engineer",
      skillIds: [],
      maxHops: 4,
    });

    expect(results.map((result) => result.hops)).toEqual([1, 2]);
  });
});

File: backend/src/test/java/com/talentgraph/service/CareerPathServiceTest.java
package com.talentgraph.service;

import com.talentgraph.model.CandidateCareerPath;
import com.talentgraph.model.CareerPathResult;
import com.talentgraph.model.RoleRequirementRow;
import com.talentgraph.model.RoleSummary;
import com.talentgraph.model.SkillRequirement;
import com.talentgraph.model.TransitionSummary;
import com.talentgraph.repository.CareerPathRepository;
import com.talentgraph.repository.LearningRepository;
import com.talentgraph.repository.RoleRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class CareerPathServiceTest {

    @Mock
    private RoleRepository roleRepository;

    @Mock
    private CareerPathRepository careerPathRepository;

    @Mock
    private LearningRepository learningRepository;

    @InjectMocks
    private CareerPathService careerPathService;

    @Test
    void ordersShorterPathsFirstAndRemovesDuplicateCandidates() {
        RoleSummary frontend = role("frontend-developer", "Frontend Developer");
        RoleSummary fullStack = role("full-stack-developer", "Full Stack Developer");
        RoleSummary aiEngineer = role("ai-engineer", "AI Engineer");

        CandidateCareerPath shortPath = new CandidateCareerPath(
                List.of(frontend, aiEngineer),
                List.of(new TransitionSummary("hard", "Build the missing AI foundation.")),
                1
        );
        CandidateCareerPath longPath = new CandidateCareerPath(
                List.of(frontend, fullStack, aiEngineer),
                List.of(
                        new TransitionSummary("moderate", "Expand into server-side delivery."),
                        new TransitionSummary("hard", "Develop machine-learning capabilities.")
                ),
                2
        );

        when(roleRepository.findById("frontend-developer")).thenReturn(frontend);
        when(roleRepository.findById("ai-engineer")).thenReturn(aiEngineer);
        when(careerPathRepository.findCandidates(
                "frontend-developer", "ai-engineer", 4
        )).thenReturn(List.of(longPath, shortPath, shortPath));
        when(roleRepository.findRequirements(anyList())).thenReturn(List.of(
                row("frontend-developer", requirement("javascript", 5)),
                row("full-stack-developer", requirement("javascript", 5)),
                row("full-stack-developer", requirement("node-js", 4)),
                row("ai-engineer", requirement("python", 5))
        ));
        when(learningRepository.findOptionsForSkills(anyList())).thenReturn(List.of());

        List<CareerPathResult> results = careerPathService.findPaths(
                "frontend-developer",
                "ai-engineer",
                List.of("javascript"),
                4
        );

        assertThat(results).extracting(CareerPathResult::hops).containsExactly(1, 2);
        assertThat(results).extracting(CareerPathResult::id).doesNotHaveDuplicates();
    }

    private static RoleRequirementRow row(String roleId, SkillRequirement requirement) {
        return new RoleRequirementRow(roleId, requirement);
    }

    private static SkillRequirement requirement(String id, int importance) {
        return new SkillRequirement(
                id,
                id,
                id,
                "Test",
                "Practical knowledge for test scenarios.",
                importance,
                "intermediate",
                true
        );
    }

    private static RoleSummary role(String id, String name) {
        return new RoleSummary(
                id,
                id,
                name,
                "Engineering",
                "mid",
                "Builds production software systems."
        );
    }
}
